Add macaroon based RPC middleware interceptor

guggero opened the pull request:
With this PR we introduce the concept of RPC middleware: A mechanism
similar to the existing channel or HTLC interceptors but this time for
gRPC messages themselves.
An RPC middleware can register itself to the main RPC server to get
notified each time a new gRPC request comes in, a gRPC response is sent
back or a streaming RPC is connected. The middleware can
validate/inspect incoming requests and modify/overwrite outgoing
responses.

The new flow for gRPC requests with this PR looks like this:

```
//        |
//        | gRPC request from client
//        |
//    +---v--------------------------------+
//    |   InterceptorChain                 |
//    +-+----------------------------------+
//      | Log Interceptor                  |
//      +----------------------------------+
//      | RPC State Interceptor            |
//      +----------------------------------+
//      | Macaroon Interceptor             |
//      +----------------------------------+--------> +---------------------+
//      | RPC Macaroon Middleware Handler  |<-------- | External Middleware |
//      +----------------------------------+          |   - approve request |
//      | Prometheus Interceptor           |          +---------------------+
//      +-+--------------------------------+
//        | validated gRPC request from client
//    +---v--------------------------------+
//    |   main gRPC server                 |
//    +---+--------------------------------+
//        |
//        | original gRPC request to client
//        |
//    +---v--------------------------------+--------> +---------------------+
//    |   RPC Macaroon Middleware Handler  |<-------- | External Middleware |
//    +---+--------------------------------+          |   - modify response |
//        |                                           +---------------------+
//        | edited gRPC request to client
//        v
```

## Motivation

This PR was initially motivated by https://github.com/lightningnetwork/lnd/issues/291. Adding all that business logic to `lnd` itself just didn't seem like the reasonable way to implement accounts.

Instead a similar mechanism to the channel and HTLC acceptors would make more sense as it would also allow other developers to implement custom business logic _outside_ of `lnd`, remotely similar to how plugins work in c-lightning for example.
The main advantage of the approach in this PR is that the custom business logic would not live in the same process (or even a sub-process) of `lnd`. The main disadvantage lies in the added request/response latency that intercepting each message can add.

An example of how this can be used to implement an account system on top of `lnd` can be seen [here, where the functionality is added to faraday](https://github.com/guggero/faraday/commit/51a7957187776a4e3e600c5993770190162b32e1).

## Security

Since this also opens the door for malicious software to interfere with
lnd in a negative way, we bind everything to macaroons with custom
caveat conditions: A middleware declares upon registration which custom
caveat name it can handle. Only client requests that send a macaroon
with that custom caveat will then be given to the middleware for
inspection. The alternative is for a middleware to register for
read-only access. In that case it will get access to all requests and
responses but isn't allowed to modify anything.
Therefore requests with the default, unencumbered macaroons
can never be modified by any middleware.

Fixes #4383.

alexbosworth commented:
How do you pass extra data to the middleware in the request? Modify the proto to include the additional data?

guggero commented:
This isn't supported with the current version, really. Modifying the proto would kind of defeat the purpose of having a generic interceptor/middleware that allows you to use a stock `lnd`.
It could be done in a number of different ways though. What kind of additional information did you have in mind?
In the use case I built this for, the only extra information required is a caveat in the macaroon.

alexbosworth commented:
> This isn't supported with the current version, really. Modifying the proto would kind of defeat the purpose of having a generic interceptor/middleware that allows you to use a stock `lnd`.
> It could be done in a number of different ways though. What kind of additional information did you have in mind?
> In the use case I built this for, the only extra information required is a caveat in the macaroon.

I don't have any specific information in mind, but since you can mutate the macaroon I guess you can add more request data there?

I'm curious why the interceptor doesn't see the mutated request data if it has access to the raw protobuf request blob?

gkrizek commented:
Would registering your middleware to the main RPC require a macaroon?

Once your middleware is registered, it only see requests from macaroons with the custom caveat?

guggero commented:
Yes, to register the middleware you need a macaroon with the `macaroon:write` permission which only the admin default macaroon has. You also tell `lnd` which custom macaroon caveat your middleware handles and lnd will then only let you intercept any requests that have a macaroon with that caveat.

gkrizek commented:
Got it. I think this is a great idea.

Do you think there would ever be a situation to allow _all_ macaroons to go through the middleware or to use a macaroon other than `macaroon:write` to register? I'll explain an idea I had;

What if I wanted to have a readonly middleware that could keep an access logs of the commands ran on the node? In that scenario it would be great to use a macaroon with `macaroon:read` permissions to register the RPC middleware. And when I use the `macaroon:read` permission to register that means the middleware could never take any action on a request. Just view it. This would also be a good case to see all macaroon actions, not just custom caveats. Just a thought I had...

guggero commented:
Yeah, I think with the mechanism in place where the middleware has to "register" itself after connecting, we could definitely add some feature toggles. Where you could say "I want read access but to all requests".

And maybe we could add a new macaroon permission as well, for example `middleware:read` and `middleware:write`. With the `read` permission you would only be allowed to set the above mentioned flag to `true` anyway.
I think that could be very useful!

Roasbeef commented on rpcperms/middleware_handler.go:
```diff
@@ -0,0 +1,448 @@
+	timeout time.Duration, params *chaincfg.Params,
+	quit chan struct{}) *MiddlewareHandler {
+
+	// We explicitly want to log this as a warning since intercepting any
```
I assume users would need to activate this config _manually_ and also obtain a special (possibly with additional locked down perms on disk) macaroon in order to interact w/ the service?

guggero replied:
Yeah, the middleware needs a macaroon with the `macaroon:write` permission to interact with the service to register itself. There currently is no global config to turn the entire feature on or off. Do you think we should add a flag? And should the feature be on or off by default?

Roasbeef replied:
> Do you think we should add a flag? And should the feature be on or off by default?

Makes sense to have it behind a flag IMO, the feature is super powerful, so I think we want to observe it a bit in the wild to see how we can make it more secure and error prone before was make it a new default feature.

Roasbeef commented on rpcperms/middleware_handler.go:
```diff
@@ -0,0 +1,448 @@
+
+				// For intercepted responses we also allow the
+				// content itself to be overwritten.
+				if requestInfo.request.Type == TypeResponse &&
```
So the idea is to allow new middleware to _modify_ a normal `lnd` response? I guess a use case for this could be adding extra account ID/detail information?

guggero replied:
Yes exactly. The middleware is restricted to the fields the proto defines. But you can filter or add information. In our faraday accounts example we would not return any channels for example or only show the user's payments to hide some of the complexity of a node to custodial users.

Roasbeef commented on rpcperms/middleware_handler.go:
```diff
@@ -0,0 +1,448 @@
+	}
+}
+
+type InterceptType uint8
```
PR still marked as WIP, but just want to mark here that we're missing `godoc` comments here.

Roasbeef commented on lnrpc/rest-annotations.yaml:
```diff
@@ -139,6 +139,8 @@ http:
     - selector: lnrpc.Lightning.ListPermissions
       get: "/v1/macaroon/permissions"
+    - selector: lnrpc.Lightning.RegisterRPCMiddleware
+      # bi-directional streaming RPC, REST not supported
```
IIRC it is now?

guggero replied:
Yeah, this PR is quite old :joy:

Roasbeef commented on rpcperms/interceptor.go:
```diff
@@ -165,6 +207,66 @@ func (r *InterceptorChain) Permissions() map[string][]bakery.Op {
+	// restricted permissions based upon the custom caveat condition)
+	// macaroon would not be validated against the limitations that the
+	// custom caveat implicate.
+	for _, middleware := range r.registeredMiddleware {
```
Given it's a map, why not just look it up here directly?

guggero replied:
Good point, fixed.

Roasbeef replied:
Still here in the latest commit?

Roasbeef reviewed (changes requested):
Woah, this is gonna be soooo powerful 

The implementation is also rather streamlined IMO, first review pass, left a few initial comments. Will also check out that Faraday example so I can get a better idea w.r.t how this can be used in practice

guggero commented on cmd/lncli/cmd_macaroon.go:
```diff
@@ -66,6 +66,14 @@ var bakeMacaroonCommand = cli.Command{
 			Name:  "ip_address",
 			Usage: "the IP address the macaroon will be bound to",
 		},
+		cli.StringFlag{
```
Hmm, thought this was a french word that also existed in English. But apparently it doesn't exist in either language :joy:

guggero commented on cmd/lncli/cmd_macaroon.go:
```diff
@@ -66,6 +66,14 @@ var bakeMacaroonCommand = cli.Command{
+			Usage: "the name of the custom caveat to add",
+		},
+		cli.StringFlag{
+			Name:  "custom_caveat_condition",
```
The custom condition here would be something like `account 12345678` to lock any request to an account. The software intercepting the requests (in my example that would be faraday) would then make sure the account mentioned (which exists in the software's own DB) has enough balance for the requested operation.

guggero commented:
Rebased and addressed commits.

The following things still need to be added before the WIP state can be removed:
 - Add unit tests
 - Add integration tests
 - Add flag to enable interceptor globally

guggero commented:
Update: only one TODO remaining, removing WIP state now:
- Add itest case for overwriting a response.

guggero commented:
Added itest case for response manipulation as well now.

Kixunil commented on config.go:
```diff
@@ -225,15 +232,19 @@ type Config struct {
+
 	AcceptorTimeout time.Duration `long:"acceptortimeout" description:"Time after which an RPCAcceptor will time out and return false if it hasn't yet received a response"`
 
+	RPCMiddleware        bool          `long:"rpcmiddleware" description:"Enable the RPC middleware interceptor functionality."`
```
I think there should also be `MandateRPCMiddleware` that will cause LND to fail all RPC requests except middleware registration if middleware is not registered. This way issues around middleware crashing or not starting would not cause security problems. I assume there exists a unique middleware macaroon. (Or should there be multiple middlewares?)

guggero replied:
That's a good idea! The current design allows for multiple middlewares to be registered at the same time. So maybe this should be a comma separated list called `MandatoryRPCMiddleware`? Requests will only be allowed if all listed middlewares are registered?

Kixunil replied:
Repeated argument/property would be slightly more preferable as it's easier to create configuration file from multiple pieces by just running `cat` but other than that (and perhaps plural name) it sounds reasonable.

guggero replied:
Added the `--middleware.addmandatory` config option/flag that can be used multiple times.

Roasbeef commented on cmd/lncli/cmd_macaroon.go:
```diff
@@ -181,6 +209,13 @@ func bakeMacaroon(ctx *cli.Context) error {
 			macaroons.IPLockConstraint(ipAddress.String()),
 		)
 	}
+	if customCaveatName != "" {
```
Ahh gotcha, thanks for the explanation

Roasbeef commented on macaroons/service.go:
```diff
@@ -265,3 +265,21 @@ func MacaroonFromContext(ctx context.Context) (*macaroon.Macaroon, error) {
+// modified. This is necessary because the macaroon library's own Clone() method
+// is unsafe for certain edge cases, resulting in both the cloned and the
+// original macaroons to be modified.
+func SafeCopyMacaroon(mac *macaroon.Macaroon) (*macaroon.Macaroon, error) {
```
👍

Roasbeef commented on rpcperms/interceptor.go:
```diff
@@ -593,3 +707,202 @@ func (r *InterceptorChain) rpcStateStreamServerInterceptor() grpc.StreamServerIn
+	r.RLock()
+	defer r.RUnlock()
+
+	for _, middleware := range r.registeredMiddleware {
```
Hmm, don't we only want to send to the middleware the precisely matches the specified custom caveat?

guggero replied:
We give each request to every middleware that qualifies. There can only be one middleware per unique custom caveat name. But there is also the concept of "read-only" middleware that can inspect _ever_ request (but can never manipulate a response). So a request can be sent out to multiple middlewares sequentially. The `if` below checks what middleware qualifies.

Roasbeef commented on rpcperms/interceptor.go:
```diff
@@ -593,3 +707,202 @@ func (r *InterceptorChain) rpcStateStreamServerInterceptor() grpc.StreamServerIn
+	defer r.RUnlock()
+
+	currentMessage := m
+	for _, middleware := range r.registeredMiddleware {
```
Similar question here re only interacting with the specified middleware.

Roasbeef commented on rpcperms/interceptor.go:
```diff
@@ -593,3 +707,202 @@ func (r *InterceptorChain) rpcStateStreamServerInterceptor() grpc.StreamServerIn
+
+// serverStreamWrapper is a struct that wraps a server stream in a way that all
+// requests and responses can be intercepted individually.
+type serverStreamWrapper struct {
```
Very cool wrapper!

Roasbeef reviewed (approved):
LGTM 🥂

Just a few clarifying questions here w.r.t the target scope of each registered middleware.

carlaKC commented on rpcserver.go:
```diff
@@ -6952,3 +6956,110 @@ func (r *rpcServer) FundingStateStep(ctx context.Context,
+			"name of at least %d characters", nameMinLength)
+	}
+
+	readOnly := registerMsg.ReadOnlyMode
```
Do we need a nil check in case the middleware doesn't send a register msg first?

guggero replied:
Yes, that's done a few lines above.

carlaKC replied:
oops, meant to delete this!

carlaKC reviewed (commented):
Dropping the first half of a review in here, really looking forward to this feature!

carlaKC commented on lnrpc/lightning.proto:
```diff
@@ -3898,3 +3898,187 @@ message Op {
+        /*
+        Intercept incoming gRPC client request message: all incoming messages,
+        both on streaming and unary RPCs, are forwarded to the middleware for
+        inspection. For unary RPC messages the middleware is also expected to
```
> "For unary RPC messages the middleware is also expected to validate the custom macaroon caveat of the request."

Wouldn't middleware also need to validate custom macaroon caveats for streaming calls?

guggero replied:
Yes. The difference is that for an unary call, the "setup" or "authorization" of an RPC call (when we "see" the call for the first time in the gRPC interceptor and want to decide whether to accept or refuse the call) and the arrival of the gRPC request message happens in the same call. That's why both the _authorization_ and the interception of the actual request message happen in the same step.
For a streaming call, the "setup" or "authorization" (decide whether to accept or refuse a stream initialization) can happen a long time before the first gRPC request message arrives (if ever). That's why it's split into two interception steps, so you can individually authorize (accept/refuse) a call, independent of whether a message is sent initially or not.
Hope that makes sense. Let me know if you think that needs to be reflected better in the comments.

carlaKC commented on lnrpc/lightning.proto:
```diff
@@ -3898,3 +3898,187 @@ message Op {
+    A boolean indicating that the gRPC response should be replaced/overwritten.
+    As its name suggests, this can only be used as a feedback to an intercepted
+    response RPC message and is ignored for feedback on any other message. This
+    boolean is needed because in protobuf an empty message is serialized as a
+    0-length or nil byte slice and we wouldn't be able to distinguish between
+    an empty replacement message and the "don't replace anything" case.
```
nice clarification!

carlaKC commented on rpcperms/middleware_handler.go:
```diff
@@ -0,0 +1,508 @@
+var (
+	// errShuttingDown is the error that's returned when the server is
+	// shutting down and a request cannot be served anymore.
+	errShuttingDown = errors.New("server shutting down")
```
Do we want to export these? Other go projects _could_ want to import them.

carlaKC commented on rpcperms/middleware_handler.go:
```diff
@@ -0,0 +1,508 @@
+				return err
+			}
+
+		// Process newly received responses from our channel acceptor,
```
s/ channel acceptor/ interceptor? there's a few of these in this commit

guggero replied:
Oops, copy/pasting exposed :sweat_smile:  Fixed the incorrect comments, thanks.

carlaKC commented on rpcserver.go:
```diff
@@ -512,6 +512,10 @@ func MainRPCServerPermissions() map[string][]bakery.Op {
 		}},
+		lnrpc.RegisterRPCMiddlewareURI: {{
+			Entity: "macaroon",
+			Action: "write",
```
One thought here is that the read only interception also needs write perms?

guggero replied:
Yeah, I thought about that as well. But because even the read-only interception can be very intrusive from a privacy standpoint, I'd be nervous if you'd be able to intercept all traffic with just the default read-only macaroon. That's why I decided to need a write permission to _register_ your middleware.

Kixunil replied:
Does read-only macaroon mean "blanket allow all readonly calls"? Perhaps it'd be better to have "benign" macaroon (go ahead and suggest a better name) that would be recommended instead of readonly for clarity. Although it's quite a footgun to keep readonly and add privacy-intrusive permission.

carlaKC replied:
> Does read-only macaroon mean "blanket allow all readonly calls"? 

Yeah it has read permissions on all [Entities](https://github.com/lightningnetwork/lnd/blob/27b416935e3fa4fe3c0232994fa2283dfcdee344/rpcserver.go#L84). 

> That's why I decided to need a write permission to register your middleware.

Makes sense, you're _writing_ middleware that'll read only.

Kixunil replied:
> Makes sense, you're *writing* middleware that'll read only.

Ah, yes, sounds good.

carlaKC commented on rpcserver.go:
```diff
@@ -6952,3 +6956,110 @@ func (r *rpcServer) FundingStateStep(ctx context.Context,
+		if err != nil {
+			select {
+			case errChan <- err:
+			case <-ctxc.Done():
```
Wondering whether we can simplify this a bit by buffering error channel and always sending a result in? That way we don't need to check `ctx.done` + `quit` again below? Could just receive on `errChan` and return non-nil errors?

Not sure if this is more simple, just spitballing.

carlaKC reviewed (commented):
Second pass, looking great! Just a few comments from me :)

Just need to review itest + do some regtest testing now, testing with faraday example best way to do so I assume?

guggero commented on macaroons/constraints.go:
```diff
@@ -87,3 +115,90 @@ func IPLockChecker() (string, checkers.Func) {
+// a macaroon.
+func CustomConstraint(name, condition string) func(*macaroon.Macaroon) error {
+	return func(mac *macaroon.Macaroon) error {
+		if name != "" {
```
Hmm, I think we should even return an error in that case, which I added now.

guggero commented on macaroons/constraints.go:
```diff
@@ -87,3 +115,90 @@ func IPLockChecker() (string, checkers.Func) {
+		// condition of the custom caveat (we add a layer of conditions
+		// to allow one custom checker to work for all custom lnd
+		// conditions that implement arbitrary business logic).
+		parts := strings.Split(outerCondition, " ")
```
No, I think we don't want to restrict what the condition should look like. It might contain some spaces itself, which is fine, as long as we can extract a name from the first part.

guggero commented on cmd/lncli/cmd_macaroon.go:
```diff
@@ -112,6 +124,24 @@ func bakeMacaroon(ctx *cli.Context) error {
 	}
 
+	if ctx.IsSet("custom_caveat_name") {
+		customCaveatName = strings.TrimSpace(ctx.String(
```
Good idea, done.

guggero commented on config.go:
```diff
@@ -225,13 +225,14 @@ type Config struct {
+	LogDir         string `long:"logdir" description:"Directory to log output."`
+	MaxLogFiles    int    `long:"maxlogfiles" description:"Maximum logfiles to keep (0 for no rotation)"`
+	MaxLogFileSize int    `long:"maxlogfilesize" description:"Maximum logfile size in MB"`
+
```
Ah, that was left over from a previous version. Good catch, fixed.

guggero commented on lncfg/rpcmiddleware.go:
```diff
@@ -0,0 +1,27 @@
+// RPCMiddleware holds the configuration for RPC interception middleware.
+type RPCMiddleware struct {
+	Enable           bool          `long:"enable" description:"Enable the RPC middleware interceptor functionality."`
+	InterceptTimeout time.Duration `long:"intercepttimeout" description:"Time after which a RPC middleware intercept request will time out and return an error if it hasn't yet received a response."`
```
Yeah, good point. Fixed.

Kixunil commented:
I wonder what would be the best way to distinguish two applications using different macaroons but essentially having same permissions. Is custom caveat condition good enough for this or is there a better way?

guggero commented:
> I wonder what would be the best way to distinguish two applications using different macaroons but essentially having same permissions. Is custom caveat condition good enough for this or is there a better way?

Currently there is no way of more fine-grained control over what you are able to do with an RPC, only whether or not you're allowed to access that RPC or not. That's why I went with the higher permission requirement for the middleware registration RPC such that you need a macaroon with _write_ permission to register a middleware, even if that middleware only registers for read-only interception.

I'm not sure how we could make that more fine-grained except for offering two separate RPCs for read-only and normal interception. But I'd rather avoid doing that.

We _could_ do some custom macaroon permission checking in the registration. So allow a middleware to register using a macaroon with only the `macaroon:read` permission and then manually check whether the macaroon also contains the `macaroon:write` permission only if it tries to register for write access.
But I'm not sure that is such a good idea, to start moving custom macaroon permission checks within the RPC methods themselves...

Kixunil commented:
My last question was not related to read/write permissions.

Basically I can see at least two use cases:

* Have multiple users with budgets, each gets a unique macaroon and budgets are tracked by the middleware.
* Let an application only see invoices it created - so if an application is compromised the privacy leak is limited.

In order to achieve this the middleware has to be able to distinguish macaroons somehow. What's the best way?

Kixunil commented:
I'm asking whether putting the client identifier into custom condition is a good way to do it or if there's a better way.

guggero commented:
> I'm asking whether putting the client identifier into custom condition is a good way to do it or if there's a better way.

I think so, yes. With that construction you can safely give the macaroon to a user and they cannot remove the restriction.
If you want multiple middlewares to be able to interact with requests of the same user, you can do that as well. For example the accounting middleware would add the custom macaroon caveat `account <acct_ID>` and another , let's say custom logging middleware would add a "marker" caveat, for example `mylogger`.
A user would then get a macaroon that has both caveats applied. So any request from that user would be intercepted by both middlewares. And the logger middleware could still extract the account ID from that other caveat, even if it didn't specifically register for it.

Roasbeef commented:
Just needs a rebase and this can land, also we likely want to land this soon to avoid rebases since it touches so much in the RPC layer.

carlaKC reviewed (approved):
tACK! What an awesome change to lnd 🥇 can't wait to see what people build out on top of this. 

Didn't test the full scope of this (had to timebox my fun), but managed to cover the following with a regtest interceptor stub: 
* Readonly interceptor
* Custom interceptor w/ modification
* Custom interceptor with + without caveats
* Mandatory interceptor enforcement

Only thing that I noticed while testing is that it's difficult to use our `lndclient` with mandatory interceptors, since we do a bunch of `GetInfo` calls etc to get started, but should be easy enough to just add middle ware registration before those calls? More of a note to self than an issue with the PR.

guggero commented:
Oops, looks like I broke something in the itests with the latest rebase. Will fix on Monday.
